packages/react-apollo-hooks/src/useQuery.ts

```diff
@@ -0,0 +1,202 @@
+import ApolloClient, {
+  OperationVariables,
+  ObservableQuery,
+  ApolloError,
+  WatchQueryOptions,
+} from 'apollo-client';
+import {ObservableQueryFields} from 'react-apollo';
+import {DocumentNode} from 'graphql-typed';
+import {useEffect, useMemo, useState} from 'react';
+
+import {AsyncQueryComponentType} from '@shopify/react-graphql';
+
+import {
+  QueryHookOptions,
+  QueryHookResult,
+  QueryDocumentResolving,
+} from './types';
+import {useApolloClient} from './ApolloContext';
+import useGraphQLDocument from './useGraphQLDocument';
+import {
+  getCachedObservableQuery,
+  invalidateCachedObservableQuery,
+} from './queryCache';
+import {compact, objToKey, actHack} from './utils';
+
+export default function useQuery<Data = any, Variables = OperationVariables>(
+  queryOrComponent: DocumentNode<Data, Variables>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent: AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryDocumentResolving<Data, Variables> | QueryHookResult<Data, Variables>;
+
+export default function useQuery<
```

Comment:
I mostly did the overriding so that if you pass in a queryDocument, you won't need to check to see if all the helper methods (like `refetch`), since I can infer the return type from what's being pass in.